akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

```diff
@@ -0,0 +1,169 @@
+
+      //seems like we must set handlers BEFORE preStart
+      setHandlers(in, out, initializing)
+
+      override def preStart(): Unit = {
+        super.preStart()
+        future.value match {
+          case Some(tr) =>
+            initializing.onFuture(tr)
+          case None =>
+            val cb = getAsyncCallback(initializing.onFuture)
+            future.onComplete(cb.invoke)(ExecutionContexts.parasitic)
+            //in case both ports are closed before future completion
+            setKeepGoing(true)
+        }
+      }
+
+      override def postStop(): Unit = {
+        if (!pr.isCompleted) {
+          pr.failure(new NeverMaterializedException())
+        }
+        super.postStop()
+      }
+
+      object initializing extends InHandler with OutHandler {
+        override def onPush(): Unit = {
+          //just leave it in the port
+          println("initialized: pushed")
+        }
+
+        var upstreamFailure = OptionVal.none[Throwable]
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          upstreamFailure = OptionVal.Some(ex)
+        }
+
+        override def onPull(): Unit = pull(in)
+
+        var downstreamCause = OptionVal.none[Throwable]
+        override def onDownstreamFinish(cause: Throwable): Unit = {
+          downstreamCause = OptionVal.Some(cause)
+        }
+
+        def onFuture(futureRes: Try[Flow[In, Out, M]]) = futureRes match {
+          case Failure(exception) =>
+            setKeepGoing(false)
+            pr.failure(new NeverMaterializedException(exception))
+            failStage(exception)
+          case Success(flow) =>
+            //materialize flow, connect to inlets, feed with potential events and set handlers
+            val initialized = new Initialized(flow)
+            setHandlers(in, out, initialized)
+            setKeepGoing(false)
+        }
+      }
+      class Initialized(fl: Flow[In, Out, M]) extends InHandler with OutHandler {
+        val subSource = new SubSourceOutlet[In](s"${FutureFlow.this}.subIn")
+        val subSink = new SubSinkInlet[Out](s"${FutureFlow.this}.subOut")
+
+        subSource.setHandler {
+          new OutHandler {
+            override def onPull(): Unit = {
+              if (!isClosed(in) && !hasBeenPulled(in)) {
+                pull(in)
+              }
+            }
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
+              if (!isClosed(in)) {
+                cancel(in, cause)
+              }
+            }
+          }
+        }
+        subSink.setHandler {
+          new InHandler {
+            override def onPush(): Unit = {
+              push(out, subSink.grab())
+            }
+
+            override def onUpstreamFinish(): Unit = {
+              complete(out)
+            }
+
+            override def onUpstreamFailure(ex: Throwable): Unit = {
+              fail(out, ex)
+            }
+          }
+        }
+        pr.complete {
```

Comment:
Turn this around and do something like

```scala
Try(Source.fromGraph(subSource.source).viaMat(fl)(Keep.right).to(subSink.sink).run()(subFusingMaterializer)) match {
  case Success(mat) =>
    pr.complete(mat) 
    ...
```

instead of completing and then matching.